File: sdks/akta-docs/typescript/src/scaffolder.ts
import fs from "node:fs/promises";
import path from "node:path";
import YAML from "yaml";
import type { AktaConfig } from "./config/schema.js";
import type { ScaffoldReport } from "./types.js";

const SECTION_DIRS = [
  "00-meta",
  "01-adrs",
  "02-business-core",
  "03-business-logic",
  "04-research",
  "05-features",
  "06-technical",
  "07-guides",
  "08-references",
  "09-explanations",
  "10-changelog",
  "99-archive",
] as const;

function slugify(name: string): string {
  return name
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-|-$/g, "") || "project";
}

export function defaultAktaConfig(projectName: string): AktaConfig {
  const slug = slugify(projectName);
  return {
    schema_version: "1.0.0",
    project: { name: projectName, slug },
    paths: {
      docs_root: "./akta-docs",
      include_globs: ["**/*.md"],
      exclude_globs: ["**/node_modules/**", "**/.git/**", "**/99-archive/**"],
    },
    linter: {
      default_severity: "error",
      rules: {
        "META-QUICK": { severity: "off" },
        "STRUCT-008": { severity: "off" },
        "CONTENT-001": { severity: "off" },
      },
    },
    scaffold: { create_readme_in_each_folder: true, gitkeep: false },
    changelog: {
      default_target: "akta-docs/10-changelog/CHANGELOG.md",
      entry_template: "compact",
      summary_max_length: 200,
    },
  };
}

export interface ScaffoldOptions {
  root: string;
  projectName: string;
  force: boolean;
  dryRun: boolean;
  createReadmeInEachFolder: boolean;
  gitkeep: boolean;
}

export async function runScaffold(
  options: ScaffoldOptions,
): Promise<ScaffoldReport> {
  const root = path.resolve(options.root);
  const configPath = path.join(root, ".akta-config.yaml");
  const cfg = defaultAktaConfig(options.projectName);
  cfg.scaffold = {
    create_readme_in_each_folder: options.createReadmeInEachFolder,
    gitkeep: options.gitkeep,
  };

  const docsRoot = path.join(root, "akta-docs");
  const created: string[] = [];

  if (!options.dryRun) {
    try {
      await fs.access(configPath);
      if (!options.force) {
        throw new Error(
          `${configPath} already exists; use --force to overwrite.`,
        );
      }
    } catch (e) {
      if ((e as NodeJS.ErrnoException).code !== "ENOENT") throw e;
    }
  }

  for (const dir of SECTION_DIRS) {
    const full = path.join(docsRoot, dir);
    if (!options.dryRun) {
      await fs.mkdir(full, { recursive: true });
    }
    created.push(path.join("akta-docs", dir));

    if (options.createReadmeInEachFolder) {
      const readme = path.join(full, "README.md");
      const today = new Date().toISOString().slice(0, 10);
      const docId = `${cfg.project.slug}.${dir.replace(/[^a-z0-9]+/g, "-")}-readme`;
      const body = `---
doc_id: ${docId}
doc_type: meta
date: ${today}
---

# ${dir}

Placeholder content for this section. Replace with architecture and business documentation aligned with OPENAKTA GEO standards.
`;
      if (!options.dryRun) {
        await fs.writeFile(readme, body, "utf8");
      }
      created.push(path.join("akta-docs", dir, "README.md"));
    }
    if (options.gitkeep) {
      const gk = path.join(full, ".gitkeep");
      if (!options.dryRun) {
        await fs.writeFile(gk, "", "utf8");
      }
    }
  }

  const changelogPath = path.join(docsRoot, "10-changelog", "CHANGELOG.md");
  const initialChangelog = `---
doc_id: ${cfg.project.slug}.changelog
doc_type: changelog
date: ${new Date().toISOString().slice(0, 10)}
---

# Changelog

<!-- akta-changelog-append -->
`;
  if (!options.dryRun) {
    await fs.writeFile(changelogPath, initialChangelog, "utf8");
  }
  created.push("akta-docs/10-changelog/CHANGELOG.md");

  const yamlText = YAML.stringify(cfg, { lineWidth: 0 });
  if (!options.dryRun) {
    await fs.writeFile(configPath, yamlText, "utf8");
  }

  return {
    root,
    docs_root: docsRoot,
    created_paths: created,
    config_path: configPath,
  };
}

File: sdks/akta-docs/python/src/akta_docs/scaffolder.py
from __future__ import annotations

import re
from datetime import date
from pathlib import Path

import yaml

from akta_docs.models import (
    AktaConfig,
    ChangelogCfg,
    LinterCfg,
    PathsCfg,
    ProjectCfg,
    RuleOptions,
    ScaffoldCfg,
)

SECTION_DIRS = (
    "00-meta",
    "01-adrs",
    "02-business-core",
    "03-business-logic",
    "04-research",
    "05-features",
    "06-technical",
    "07-guides",
    "08-references",
    "09-explanations",
    "10-changelog",
    "99-archive",
)


def slugify(name: str) -> str:
    s = re.sub(r"[^a-z0-9]+", "-", name.lower()).strip("-")
    return s or "project"


def default_akta_config(project_name: str) -> AktaConfig:
    slug = slugify(project_name)
    return AktaConfig(
        schema_version="1.0.0",
        project=ProjectCfg(name=project_name, slug=slug),
        paths=PathsCfg(docs_root="./akta-docs"),
        linter=LinterCfg(
            default_severity="error",
            rules={
                "META-QUICK": RuleOptions(severity="off"),
                "STRUCT-008": RuleOptions(severity="off"),
                "CONTENT-001": RuleOptions(severity="off"),
            },
        ),
        scaffold=ScaffoldCfg(),
        changelog=ChangelogCfg(
            default_target="akta-docs/10-changelog/CHANGELOG.md",
        ),
    )


class ScaffoldReport:
    def __init__(
        self,
        root: str,
        docs_root: str,
        created_paths: list[str],
        config_path: str,
    ) -> None:
        self.root = root
        self.docs_root = docs_root
        self.created_paths = created_paths
        self.config_path = config_path


def run_scaffold(
    root: str,
    project_name: str,
    *,
    force: bool = False,
    dry_run: bool = False,
    create_readme_in_each_folder: bool = True,
    gitkeep: bool = False,
) -> ScaffoldReport:
    root_p = Path(root).resolve()
    config_path = root_p / ".akta-config.yaml"
    cfg = default_akta_config(project_name)
    cfg.scaffold = ScaffoldCfg(
        create_readme_in_each_folder=create_readme_in_each_folder,
        gitkeep=gitkeep,
    )
    docs_root = root_p / "akta-docs"
    created: list[str] = []

    if not dry_run and config_path.exists() and not force:
        raise FileExistsError(f"{config_path} already exists; use --force to overwrite.")

    slug = cfg.project.slug or slugify(project_name)
    today = date.today().isoformat()

    for dir_name in SECTION_DIRS:
        full = docs_root / dir_name
        if not dry_run:
            full.mkdir(parents=True, exist_ok=True)
        created.append(str(Path("akta-docs") / dir_name))

        if create_readme_in_each_folder:
            readme = full / "README.md"
            doc_id = f"{slug}.{re.sub(r'[^a-z0-9]+', '-', dir_name)}-readme"
            body = f"""---
doc_id: {doc_id}
doc_type: meta
date: {today}
---

# {dir_name}

Placeholder content for this section. Replace with architecture and business documentation aligned with OPENAKTA GEO standards.
"""
            if not dry_run:
                readme.write_text(body, encoding="utf-8")
            created.append(str(Path("akta-docs") / dir_name / "README.md"))

        if gitkeep:
            gk = full / ".gitkeep"
            if not dry_run:
                gk.write_text("", encoding="utf-8")

    changelog_path = docs_root / "10-changelog" / "CHANGELOG.md"
    initial = f"""---
doc_id: {slug}.changelog
doc_type: changelog
date: {today}
---

# Changelog

<!-- akta-changelog-append -->
"""
    if not dry_run:
        changelog_path.write_text(initial, encoding="utf-8")
    created.append("akta-docs/10-changelog/CHANGELOG.md")

    yaml_text = yaml.safe_dump(
        cfg.model_dump(mode="json", exclude_none=True),
        sort_keys=False,
        allow_unicode=True,
    )
    if not dry_run:
        config_path.write_text(yaml_text, encoding="utf-8")

    return ScaffoldReport(
        root=str(root_p),
        docs_root=str(docs_root),
        created_paths=created,
        config_path=str(config_path),
    )
